internal/cpu.go

```diff
@@ -0,0 +1,64 @@
+package internal
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+var sysCPU struct {
+	once sync.Once
+	err  error
+	num  int
+}
+
+// PossibleCPUs returns the max number of CPUs a system may possibly have
+// Logical CPU numbers must be of the form 0-n
+func PossibleCPUs() (int, error) {
+	sysCPU.once.Do(func() {
+		sysCPU.num, sysCPU.err = parseCPUs("/sys/devices/system/cpu/possible")
+	})
+
+	return sysCPU.num, sysCPU.err
+}
+
+var onlineCPU struct {
+	once sync.Once
+	err  error
+	num  int
+}
+
+// OnlineCPUs returns the number of currently online CPUs
+// Logical CPU numbers must be of the form 0-n
+func OnlineCPUs() (int, error) {
```

Comment:
Oh, and this is currently structured in the code to be hidden from users, but in Cilium we actually report the number of CPUs that we configure perf readers for, mainly I think just as a basic sanity check to ensure that the implementation is working as intended. At the moment we don't have any way to introspect the number used by the perf library, unless we add a dependency on this `github.com/cilium/ebpf/internal`.

Not sure what the best path forward on this is. In some sense, if the library works correctly on all platforms, everything is fine. But if it doesn't work properly, by misconfiguring the number of CPUs for instance, then I'm not sure how we would introspect this. Just using the actual BPF map state for the events map?

Reply:
We have counters for failed `perf_event_output` that are exported from the eBPF, which covers this failure case. Would that work for you as well?

Alternatively we could add `Reader.NumCPU() int` or something.

Reply:
The simplest to satisfy the API we expose right now is just to get `Reader.NumCPU()`, as we literally just print the number of CPUs configured with perf events. 

We have this other problem though, which is that our C template defines the perf event map using `__NR_CPUS__` to configure the size of the map.. then from the userspace we open this map. Right now I'm doing a  `ebpf.LoadPinnedMapExplicit(path, eventsMapTemplate)` where the `eventsMapTemplate` is configured to have `MaxEntries: uint32(runtime.NumCPU())`, and hope that this matches (which it seems to at least on my laptop). That said, we generate our C code templates so we can always change the `__NR_CPUS__` definition to instead be written as a constant value from our golang code, and define it also using `runtime.NumCPU()`. OK, so that problem is basically solved...

So let's say we have `Reader.NumCPU()`. We still need to define the map's `MaxEntries` somehow, and this should be configured properly before we pass that map into `perf.NewReader()`. So this means it's actually a bit weird for the Reader to go and determine the number of CPUs because it will use its own method for this which is not guaranteed to be the same as the method that the calling library uses for defining the map size. How do you expect library users to determine the size of the map and what kind of assumptions is the `PerfReader()` making here based on the map size?

> We have counters for failed `perf_event_output` that are exported from the eBPF, which covers this failure case. Would that work for you as well?

This is if the BPF code attempts to output but there's no ring fd defined, right? That's also interesting, how is this exposed from the kernel? I can't recall seeing somewhere that we expose this right now but it sounds like something that could be handy to know about when weird corner case bugs come up.

Reply:
> We have this other problem though, which is that our C template defines the perf event map using __NR_CPUS__ to configure the size of the map.

I made this quick test:

```
$ cat cpu.go
package main

import (
        "fmt"
        "runtime"
)

func main() {
        fmt.Println("CPUs:", runtime.NumCPU())
}
$ go run cpu.go
CPUs: 8
$ taskset -c 0 go run cpu.go
CPUs: 1
```

Using `NumCPU` is a bug I think.

> How do you expect library users to determine the size of the map [...]

The library forces PerfEventArrays to have MaxElems == 0, and then plugs in the correct value at creation time: https://github.com/cilium/ebpf/blob/master/map.go#L100 That has the same effect as NR_CPUS but is a bit more genera, since code that creates maps directly also works. To support migration we can remove the requirement of MaxElems == 0.

There is actually another problem: the map is created with MaxElems = possible CPUs, because some systems can hotplug CPUs (@arthurfabre has one). The eBPF will then happily run on those new CPUs.  However, the perf rings are created for online CPUs, since it's impossible to create perf rings for offline CPUs. I'll create a bug for this and come back to it later. Most likely we should just declare that we don't support hotplug right now.

> [...] what kind of assumptions is the PerfReader() making here based on the map size?

Currently, it assumes that the map size == online cpus. I think you have a point, it should be possible to change that to use the size of the map instead. That also circumvents the need for `Reader.NumCPU` since you can print map.ABI().MaxElems.

> This is if the BPF code attempts to output but there's no ring fd defined, right? That's also interesting, how is this exposed from the kernel?

Yes, we just bump a counter in a per CPU array whenever `perf_event_output` returns < 0. When collecting metrics we sum the per CPU counters into a single value, and expose that via Prometheus. This gives us very cheap metrics.

Reply:
> Using `NumCPU` is a bug I think.

Great example, agree it doesn't make sense to allocate map size based on the golang process CPU affinity.

> The library forces PerfEventArrays to have MaxElems == 0, and then plugs in the correct value at creation time: https://github.com/cilium/ebpf/blob/master/map.go#L100 [...]

For the `NewMap()` case, but not the `LoadPinnedMap()` / `LoadPinnedMapExplicit()` cases. In those cases, whoever defines the map still needs to know about and configure the map size, but then there's a possibility for this to be out-of-sync with the Reader that is opened on top of the map. If they're in sync, everything is fine. If they somehow determine the `MaxEntries` / number of CPUs differently then opening the reader will consistently fail.

I guess where I'm going with this is, should we consider making the `PossibleCPUs` and/or `OnlineCPUs` part of the library API which users could then use to configure their maps for the pinned case? This would be useful not only for perf event reader maps, but also PerCPU maps as well.

> Most likely we should just declare that we don't support hotplug right now.

:+1: 

> Yes, we just bump a counter in a per CPU array [...]

Ah, makes sense. So there's not really a generic way to integrate this into the library, for example as statistics that could be fetched via some perf API. Anyway, sounds useful, thanks for the explanation.

Reply:
> I guess where I'm going with this is, should we consider making the PossibleCPUs and/or OnlineCPUs part of the library API which users could then use to configure their maps for the pinned case? This would be useful not only for perf event reader maps, but also PerCPU maps as well.

I was wondering if I'm thinking too narrowly on this, why not create the map from Cilium agent side? However, we still have the upgrade case where we would presumably prefer to leave the existing map in place, as-is, and not recreate it every time we upgrade or restart Cilium. So then we still need the ability to open the map from a pinned location on the filesystem, and (given our support for <4.13 kernels) will need to provide the map size in the ABI for the `LoadPinnedMapExplicit()` call.

Reply:
Once we have #5 `LoadPinnedMap` will work on earlier kernels, so you won't need to use the explicit version. Which means you don't need to know the number of CPUs.